File: app/src/main/java/com/cookplan/shopping_list/list_by_dishes/ShopListByDishesFragment.java
package com.cookplan.shopping_list.list_by_dishes;

import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ExpandableListView;
import android.widget.ProgressBar;

import com.cookplan.BaseFragment;
import com.cookplan.R;
import com.cookplan.models.Ingredient;
import com.cookplan.models.Recipe;
import com.cookplan.models.ShopListStatus;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


public class ShopListByDishesFragment extends BaseFragment implements ShopListByDishesView {

    private ShopListByDishPresenter presenter;
    private ProgressBar progressBar;

    public ShopListByDishesFragment() {
    }

    public static ShopListByDishesFragment newInstance() {
        ShopListByDishesFragment fragment = new ShopListByDishesFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
        }
        setRetainInstance(true);
        presenter = new ShopListByDishesPresenterImpl(this, getActivity());
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        mainView = (ViewGroup) inflater.inflate(R.layout.fragment_shop_list_by_dish, container, false);

        progressBar = (ProgressBar) mainView.findViewById(R.id.progress_bar);
        return mainView;
    }

    @Override
    public void onStart() {
        super.onStart();
        if (presenter != null) {
            presenter.getShoppingList();
        }
        setEmptyView();
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void onStop() {
        super.onStop();
        if (presenter != null) {
            presenter.onStop();
        }
    }

    @Override
    public void setErrorToast(String error) {
        Snackbar.make(mainView, getString(R.string.error_load_shop_list), Snackbar.LENGTH_LONG).show();
    }

    @Override
    public void setEmptyView() {
        progressBar.setVisibility(View.GONE);
        setEmptyViewVisability(View.VISIBLE);
        setContentVisability(View.GONE);
    }

    private void setContentVisability(int visability) {
        ViewGroup contentLayout = (ViewGroup) mainView.findViewById(R.id.main_content_layout);
        if (contentLayout != null) {
            contentLayout.setVisibility(visability);
        }
    }

    @Override
    public List<Recipe> getExistedRecipeList() {
        ExpandableListView expandableListView = (ExpandableListView) mainView.findViewById(R.id.shop_list_by_dish_expListView);
        ShopListExpandableListAdapter adapter = (ShopListExpandableListAdapter) expandableListView.getExpandableListAdapter();
        return adapter != null ? adapter.getRecipeList() : new ArrayList<>();
    }

    @Override
    public Map<String, List<Ingredient>> getExistedRecipeIdsToingredientsMap() {
        ExpandableListView expandableListView = (ExpandableListView) mainView.findViewById(R.id.shop_list_by_dish_expListView);
        ShopListExpandableListAdapter adapter = (ShopListExpandableListAdapter) expandableListView.getExpandableListAdapter();
        return adapter != null ? adapter.getRecipeIdsToIngredientMap() : new HashMap<>();
    }

    @Override
    public void setIngredientListToRecipe(List<Recipe> newGroupList, Map<String, List<Ingredient>> newChildMap) {
        progressBar.setVisibility(View.GONE);
        setEmptyViewVisability(View.GONE);
        setContentVisability(View.VISIBLE);

        if (getActivity() != null) {
            ExpandableListView expandableListView = (ExpandableListView) mainView.findViewById(R.id.shop_list_by_dish_expListView);
            ShopListExpandableListAdapter needToBuyAdapter = new ShopListExpandableListAdapter(
                    getActivity(),
                    newGroupList,
                    newChildMap,
                    new ShopListExpandableListAdapter.OnItemClickListener() {
                        @Override
                        public void onIngredientClick(Ingredient ingredient) {
                            ShopListStatus newStatus;
                            if (ingredient.getShopListStatus() == ShopListStatus.NEED_TO_BUY) {
                                newStatus = ShopListStatus.ALREADY_BOUGHT;
                            } else {
                                newStatus = ShopListStatus.NEED_TO_BUY;
                            }
                            if (presenter != null) {
                                presenter.setIngredientBought(ingredient, newStatus);
                            }
                        }

                        @Override
                        public void onDeleteGroupClick(Recipe recipe, List<Ingredient> ingredientList) {
                            new AlertDialog.Builder(getActivity(), R.style.AppCompatAlertDialogStyle).setTitle(R.string.attention_title)
                                    .setMessage(R.string.delete_recipe_from_shop_list_question)
                                    .setPositiveButton(android.R.string.ok, (dialog, which) -> {
                                        if (presenter != null) {
                                            presenter.setRecipeIngredBought(recipe, ingredientList);
                                        }
                                    })
                                    .setNegativeButton(android.R.string.no, null)
                                    .show();

                        }
                    });
            expandableListView.setAdapter(needToBuyAdapter);
            int count = needToBuyAdapter.getGroupCount();
            for (int position = 0; position < count; position++) {
                expandableListView.expandGroup(position);
            }
        }
    }
}

File: app/src/main/java/com/cookplan/shopping_list/list_by_dishes/ShopListByDishesFragment.kt
package com.cookplan.shopping_list.list_by_dishes

import android.content.Context
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.cookplan.BaseFragment
import com.cookplan.R
import com.cookplan.models.Ingredient
import com.cookplan.models.Recipe


class ShopListByDishesFragment : BaseFragment(), ShopListByDishesView {


    private var presenter: ShopListByDishPresenter? = null
    private var progressBarLayout: View? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
        presenter = ShopListByDishesPresenterImpl(this, activity as Context)
    }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        mainView = inflater.inflate(R.layout.fragment_shop_list_by_dish, container, false) as ViewGroup
        progressBarLayout = mainView?.findViewById(R.id.progress_bar_layout)
        return mainView
    }

    override fun onStart() {
        super.onStart()
        presenter?.getShoppingList()
        setEmptyView()
        progressBarLayout?.visibility = View.VISIBLE
    }

    override fun onStop() {
        super.onStop()
        presenter?.onStop()
    }

    override fun setErrorToast(error: String) {
        super.setErrorToast(getString(R.string.error_load_shop_list));
    }

    override fun setEmptyView() {
        progressBarLayout?.visibility = View.GONE
        setEmptyViewVisability(View.VISIBLE)
        setContentVisability(View.GONE)
    }

    private fun setContentVisability(visability: Int) {
        val contentLayout = mainView?.findViewById<ViewGroup>(R.id.main_content_layout)
        contentLayout?.visibility = visability
    }

    override fun setIngredientListToRecipe(newGroupList: MutableList<Recipe>, newChildMap: Map<String, List<Ingredient>>) {
        progressBarLayout?.visibility = View.GONE

        if (newGroupList.isEmpty()) {
            setEmptyView()
        } else {
            setEmptyViewVisability(View.GONE)
            setContentVisability(View.VISIBLE)
            val needToBuyRecyclerView = mainView?.findViewById<RecyclerView>(R.id.shop_list_by_dish_recycler)
            needToBuyRecyclerView?.setHasFixedSize(true)
            needToBuyRecyclerView?.isNestedScrollingEnabled = false
            needToBuyRecyclerView?.layoutManager = LinearLayoutManager(activity)

            needToBuyRecyclerView?.adapter = ShopListByDishesRecyclerAdapter(newGroupList, newChildMap,
                    { recipe, ingredientList ->
                        AlertDialog.Builder(activity as Context, R.style.AppCompatAlertDialogStyle).setTitle(R.string.attention_title)
                                .setMessage(R.string.delete_recipe_from_shop_list_question)
                                .setPositiveButton(android.R.string.ok) { dialog, which ->
                                    progressBarLayout?.visibility = View.VISIBLE
                                    presenter?.recipeIngredBought(recipe, ingredientList)
                                }
                                .setNegativeButton(android.R.string.no, null)
                                .show()
                    })

//            val expandableListView = mainView?.findViewById<View>(R.id.shop_list_by_dish_expListView) as ExpandableListView
//            val needToBuyAdapter = ShopListExpandableListAdapter(
//                    activity as Context,
//                    newGroupList,
//                    newChildMap,
//                    { ingredient ->
//                        val newStatus: ShopListStatus
//                        if (ingredient.shopListStatus === ShopListStatus.NEED_TO_BUY) {
//                            newStatus = ShopListStatus.ALREADY_BOUGHT
//                        } else {
//                            newStatus = ShopListStatus.NEED_TO_BUY
//                        }
//                        presenter?.changeIngredientStatus(ingredient, newStatus)
//
//                    }, )
//            expandableListView.setAdapter(needToBuyAdapter)
//            val count = needToBuyAdapter.groupCount
//            for (position in 0 until count) {
//                expandableListView.expandGroup(position)
//            }
        }
    }

    companion object {

        @JvmStatic
        fun newInstance(): ShopListByDishesFragment {
            val fragment = ShopListByDishesFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }
}